notebook/notebookapp.py

```diff
@@ -1277,7 +1277,9 @@ def _confirm_exit(self):
             line = sys.stdin.readline()
             if line.lower().startswith(yes) and no not in line.lower():
                 self.log.critical(_("Shutdown confirmed"))
-                ioloop.IOLoop.current().stop()
+                # schedule stop on the main thread,
+                # since this might be called from a signal handler
```

Comment:
The docs for `add_callback` say that it shouldn't be used from a signal handler - there's a separate method for that.

http://www.tornadoweb.org/en/stable/ioloop.html#tornado.ioloop.IOLoop.add_callback

Reply:
Thanks. Fixed.

Reply:
Sorry, I've just gone and looked at the code, and we actually call this in a thread started by a signal handler. I'm not 100% sure, but I think that should be using the `add_callback()` method after all.